feat: vue starter app

daine opened the pull request:
# Summary | Résumé
Adds a starter app for usage in Vue 3.

This starter app contains the following:
- Regular JS (I'm hoping we can produce a typescript equivalent in the near future)
- Vue Router
- Pinia for state management
- VI tests for unit tests
- Playwright for E2E tests
- ESLint configuration
- Prettier configuration

daine commented:
There is currently a dependency issue with send version 0.18.0 which I've traced to `serve-static`. [Here is a PR](https://github.com/expressjs/serve-static/pull/176), hoping it gets merged and published soon.

melaniebmn commented on starter-apps/vue/vue-template/src/components/Header.vue:
```diff
@@ -0,0 +1,92 @@
+import AppLink from '@/components/AppLink.vue'
+import { getLocalizedPath } from '@/router/index.js'
+import { getOtherLocale } from '@/i18n/index.js'
+import HeaderBreadcrumbs from '@/components/HeaderBreadcrumbs.vue'
```
Could we rename this component from `HeaderBreadcrumbs` to simply `Breadcrumbs`? Since breadcrumbs are intended to be used only in the header, including "header" in the component name seems redundant.

daine replied:
This is the typical naming convention in vue - I've had to add single-word components as exceptions in the eslint file so it ignores them. It is preferred to name things this way especially if they are nested within that component, in this case, `Breadcrumbs` are only found inside of `Header`, hence `HeaderBreadcrumbs`.

More info here: https://vuejs.org/style-guide/rules-strongly-recommended.html#base-component-names

melaniebmn commented on starter-apps/vue/vue-template/src/components/Header.vue:
```diff
@@ -0,0 +1,92 @@
+</script>
+
+<template>
+  <gcds-header skip-to-href="#main">
```
I believe this should be updated to `skip-to-href="#main-content", as the main container uses the ID `main-content`.

melaniebmn commented on starter-apps/vue/vue-template/src/components/Header.vue:
```diff
@@ -0,0 +1,92 @@
+    <gcds-search slot="search"></gcds-search>
+    <div slot="toggle">
+      <AppLink :to="getOtherLangPath">{{ t('langToggle') }}</AppLink>
+    </div>
```
I believe the `toggle` slot can be replaced by using the `lang-href` prop directly on the header. To my knowledge, the slot is only available for users who prefer not to use our built-in language toggle.

daine replied:
This uses the `AppLink` component for the toggle, if I remove the toggle slot, it will behave like a regular link and won't function like a single page application

melaniebmn commented on starter-apps/vue/vue-template/src/components/Header.vue:
```diff
@@ -0,0 +1,92 @@
+</script>
+
+<template>
+  <gcds-header skip-to-href="#main">
```
Add  `lang-href` here with the correct path for each language.

daine replied:
I added it but I'm still using the `toggle` slot to take advantage of vue-router, I wonder if it still serves a purpose in this case?

melaniebmn commented on starter-apps/vue/vue-template/src/components/Header.vue:
```diff
@@ -0,0 +1,92 @@
+    <div slot="toggle">
+      <AppLink :to="getOtherLangPath">{{ t('langToggle') }}</AppLink>
+    </div>
+    <gcds-top-nav slot="menu" alignment="right" label="Top navigation">
```
You should be able to remove `alignment="right"` as that is the default value.

daine replied:
I tried it but it moved my nav items to the left :( I've put it back for now, but let me know what else I can do about it

melaniebmn commented on starter-apps/vue/vue-template/src/components/Heading.vue:
```diff
@@ -0,0 +1,6 @@
+<script setup>
+const { tag } = defineProps(['tag'])
```
Flagging to talk to you about this @daine :)

melaniebmn commented on starter-apps/vue/vue-template/src/components/Heading.vue:
```diff
@@ -0,0 +1,6 @@
+<script setup>
+const { tag } = defineProps(['tag'])
```
We should also add the two `margin` props.

daine replied:
Those two other props are optional which is why this didn't automatically include them -- but I can include that here

melaniebmn commented on starter-apps/vue/vue-template/src/components/forms/Input.vue:
```diff
@@ -0,0 +1,20 @@
+  'label',
+  'placeholder',
+  'required'
+])
```
This seems to be missing some important props like `hint` and `error-message` for example.

daine replied:
I've gotta be honest -- the biggest reason they're not here by default is because I only initially included required props. Are those props important enough that we need to make them required?

melaniebmn commented on starter-apps/vue/vue-template/src/components/forms/TextArea.vue:
```diff
@@ -0,0 +1,19 @@
+  'label',
+  'placeholder',
+  'required'
+])
```
Same comment: This seems to be missing some important props like `hint` and `error-message` for example.

daine replied:
Same comment as above

melaniebmn commented on starter-apps/vue/vue-template/src/views/HomeView.vue:
```diff
@@ -0,0 +1,23 @@
+      <Text>
+        {{ t('homePage.paragraph3') }}
+      </Text>
+    </section>
```
Please include the DateModified component here.

melaniebmn commented on starter-apps/vue/vue-template/src/views/About/AboutView.vue:
```diff
@@ -0,0 +1,25 @@
+    </section>
+    <section>
+      <AppLink :to="getLocalizedPath('about/about2')">{{ t('aboutPage.about2.heading') }}</AppLink>
+    </section>
```
Please include the DateModified component here.

melaniebmn commented on starter-apps/vue/vue-template/src/views/About/About1.vue:
```diff
@@ -0,0 +1,17 @@
+    <Heading class="mb-400" tag="h1">{{ t('aboutPage.about1.heading') }}</Heading>
+    <Text>
+      {{ t('aboutPage.about1.intro') }}
+    </Text>
```
Please include the DateModified component here.

melaniebmn commented on starter-apps/vue/vue-template/src/views/NotFound.vue:
```diff
@@ -0,0 +1,28 @@
+    <Heading class="mb-400" tag="h1">{{ t('notFoundPage.heading') }}</Heading>
+    <Text>
+      {{ t('notFoundPage.intro') }}
+    </Text>
```
Please include the DateModified component here.

melaniebmn commented on starter-apps/vue/vue-template/src/views/ReportABug.vue:
```diff
@@ -0,0 +1,194 @@
+          t('reportABugPage.openGithubButton')
+        }}</Button>
+      </section>
+    </div>
```
Please include the DateModified component here.

melaniebmn commented on starter-apps/vue/vue-template/src/views/HomeView.vue:
```diff
@@ -0,0 +1,23 @@
+<template>
+  <div>
+    <section>
+      <Heading class="mb-400" tag="h1">{{ t('homePage.heading') }}</Heading>
```
The heading component should automatically set the `margin-bottom` to `400`, so there is no need to apply a utility class for this.

melaniebmn commented on starter-apps/vue/vue-template/src/views/NotFound.vue:
```diff
@@ -0,0 +1,28 @@
+</script>
+<template>
+  <div class="not-found">
+    <Heading class="mb-400" tag="h1">{{ t('notFoundPage.heading') }}</Heading>
```
Same comment about `mb-400` class.

melaniebmn commented on starter-apps/vue/vue-template/src/views/ReportABug.vue:
```diff
@@ -0,0 +1,194 @@
+<template>
+  <div>
+    <section>
+      <Heading class="mb-400" tag="h1">{{ t('reportABugPage.heading') }}</Heading>
```
Same comment about mb-400 class.

melaniebmn commented on starter-apps/vue/vue-template/src/views/About/AboutView.vue:
```diff
@@ -0,0 +1,25 @@
+</script>
+<template>
+  <div class="about">
+    <Heading class="mb-400" tag="h1">{{ t('aboutPage.heading') }}</Heading>
```
Same comment about mb-400 class.

melaniebmn commented on starter-apps/vue/vue-template/src/views/About/About1.vue:
```diff
@@ -0,0 +1,17 @@
+</script>
+<template>
+  <div class="about">
+    <Heading class="mb-400" tag="h1">{{ t('aboutPage.about1.heading') }}</Heading>
```
Same comment about mb-400 class.

melaniebmn commented on starter-apps/vue/vue-template/src/views/About/AboutView.vue:
```diff
@@ -0,0 +1,25 @@
+      <AppLink :to="getLocalizedPath('about/about1')">{{ t('aboutPage.about1.heading') }}</AppLink>
+    </section>
+    <section>
+      <AppLink :to="getLocalizedPath('about/about2')">{{ t('aboutPage.about2.heading') }}</AppLink>
```
Do you think it might be beneficial to rename these pages to avoid confusion? Initially, I found it a little unclear having 3 pages with almost identical names.

daine replied:
Yes! Any recommendations? I'm feeling less creative so put these in as a placeholder 👯

ethanWallace replied:
topic-one and topic-two?

melaniebmn commented on starter-apps/vue/vue-template/src/views/About/AboutView.vue:
```diff
@@ -0,0 +1,25 @@
+      <AppLink :to="getLocalizedPath('about/about1')">{{ t('aboutPage.about1.heading') }}</AppLink>
+    </section>
+    <section>
+      <AppLink :to="getLocalizedPath('about/about2')">{{ t('aboutPage.about2.heading') }}</AppLink>
```
The About2 page appears to still be missing.

daine replied:
I put it in with the intention of it to go to a 404 (fun suprise!) but I'll remove it, it just provides confusion.

melaniebmn commented on starter-apps/vue/vue-template/src/views/About/About1.vue:
```diff
@@ -0,0 +1,17 @@
+<script lang="ts" setup>
+import { useI18n } from 'vue-i18n'
+import Heading from '@/components/Heading.vue'
+import Text from '@/components/Text.vue'
```
If we export all internal components in an index file, let's also update these imports accordingly.

melaniebmn commented on starter-apps/vue/vue-template/src/views/ReportABug.vue:
```diff
@@ -0,0 +1,194 @@
+        v-model="formData.version"
+        :label="t('reportABugPage.form.versionNumber')"
+        :name="formElements.version"
+        required
```
Could we please add hint text to all of the form fields? Following our own guidance, which recommends using hint text, would be beneficial in this case to show users the recommended usage.

melaniebmn reviewed (changes requested):
First round of review - I have a few files left that I will continue with tomorrow

ethanWallace commented on starter-apps/vue/vue-template/src/views/ReportABug.vue:
```diff
@@ -0,0 +1,204 @@
+      @gcdsValid="handleValid"
+      @submit.prevent="handleSubmit"
+    >
+      <Input
```
Could we add an error summary into this form since it is a bigger form? You could then switch each form component to validate on submit.

ethanWallace commented on starter-apps/vue/vue-template/src/components/Header.vue:
```diff
@@ -0,0 +1,92 @@
+  <gcds-header :lang-href="getOtherLangPath" skip-to-href="#main">
+    <gcds-search slot="search"></gcds-search>
+    <div slot="toggle">
+      <AppLink :to="getOtherLangPath">{{ t('langToggle') }}</AppLink>
```
Are you able to use the same logic you used for the `gcds-nav-link` on the `gcds-lang-toggle`? That way the language toggle will have the proper responsive styling.

ethanWallace reviewed (changes requested):
Just a couple comments/questions.

Do we plan to replace the lorem ipsum in the app content with other content?

daine commented:
**Updates since last review:**

- Removed font import
- Renamed package name
- Moved DateModified to each individual page
- skip-to-href id updated to reflect correct id (main-content)
- Replaced toggle slot with lang-toggle component, placed within AppLink
- Added more props to Heading: marginTop and marginBottom
- Added hint property to Input and Text Area
- Added hint text to all inputs and text area on the ReportABug form
- Added property `error-message` to the Input, but I didn't set it on the ReportABug form because it behaves a bit differently


- Renamed "HomeView" to just "Home" since it's under the `/views` folder
- Removed all added `mb-400` on unnecessary Heading component


- Rename "About1" to "Topic", completely dropped "About 2" since I don't think it was useful
- Removed invalid link to non-existent `About2` page


- Replaced Lorem Ipsum with new English text
- Added an error summary to the form


**Did not do:**
- Rename HeaderBreadcrumbs - because it is a common pattern in Vue
- On TopNav, remove alignment="right" - because it moved my nav items to the left :( 
- Add error-message to inputs on ReportABug form - its behaviour was a bit unexpected for me so we can talk about it
- Export internal components into an index file - it may be an opinionated pattern so let's leave them to decide that on their own

ethanWallace commented on starter-apps/vue/vue-template/src/views/ReportABug.vue:
```diff
@@ -0,0 +1,214 @@
+import TextArea from '@/components/forms/TextArea.vue'
+import Button from '@/components/forms/Button.vue'
+import DateModified from '@/components/DateModified.vue'
+import ErrorSummary from '@/components/forms/ErrorSummary.vue'
```
The `ErrorSummary.vue` file seems to be missing from `components/forms`.

ethanWallace reviewed (changes requested):
Just one small thing on the report a bug page. Would you also be able to update the gcds package to the latest version?

melaniebmn commented on starter-apps/vue/vue-template/package.json:
```diff
@@ -0,0 +1,37 @@
+{
+  "name": "gcds-vue-start-app",
```
Sorry nit-picky but I wanted to double-check: is it correct that we are using `start-app` here or should it be `starter-app`?

daine replied:
No prob, I was wondering about that too but you mentioned `gcds-vue-start-app` [here](https://github.com/cds-snc/gcds-examples/pull/21#discussion_r1756003231) and I simply copied it lol!

melaniebmn replied:
Haha oops, sorry about that typo!

ethanWallace reviewed (dismissed):
LGTM!

I think we could do the content revision and French content in a new PR

melaniebmn reviewed (dismissed):
Just a couple of small changes :)

daine commented on starter-apps/vue/vue-template/index.html:
```diff
@@ -0,0 +1,22 @@
+
+    <!-- GCDS required assets -->
+    <link href="https://fonts.googleapis.com" rel="preconnect" />
+    <link crossorigin href="https://fonts.gstatic.com" rel="preconnect" />
```
Done!

daine commented on starter-apps/vue/vue-template/src/components/Header.vue:
```diff
@@ -0,0 +1,91 @@
+</script>
+
+<template>
+  <gcds-header :lang-href="getOtherLangPath" skip-to-href="#main">
```
Done!

daine commented on starter-apps/vue/vue-template/src/components/HeaderBreadcrumbs.vue:
```diff
@@ -0,0 +1,48 @@
+})
+
+// TODO: Here is an example of a quick way to get breadcrumbs for each route. You may want to change this to be more dynamic.
+// In this example, only about/aboutTopic has a parent other than home.
```
```suggestion
// In this example, only about/topic has a parent other than home.
```

melaniebmn reviewed (approved):
LGTM, let's ship it!
